freqtrade/freqtradebot.py

```diff
@@ -394,21 +394,21 @@ def create_trade(self, pair: str) -> bool:
             logger.info(f"Pair {pair} is currently locked.")
             return False
 
+        if not self.get_free_open_trades():
+            logger.debug(f"Can't open a new trade for {pair}: max number of trades is reached.")
+            return False
+
+        stake_amount = self.get_trade_stake_amount(pair)
+        if not stake_amount:
+            logger.debug(f"Stake amount is 0, ignoring possible trade for {pair}.")
+            return False
+
```

Comment:
Do we really need this optimization?
I understand, this optimizes calculations (`get_signal()` won't be called till free trade slot will become available), but do we really need this change? Imagine, custom `populate_indicators()` implements some custom actions (like fetching and aggregating data from an external source) that should be done periodically...

Reply:
@hroff-1902 Actually, that's kinda precisely how I detected it, I don't see why it's needed to be this way though, but feel free to give me an example. If we want to add any functionality to the strategy, we can always support it via the interface, but this here kinda felt random.

Reply:
What exactly did you notice in `populate_*()` methods?

I struggle to see how checking this afterwards can lead to problems within `populate_*()` methods.

----

I agree with @hroff-1902 - i don't think we need this "optimization" (I wouldn't call it optimization though).
This call (`get_trade_stake_amount()`) is not a "cheap one" ... it calls `refresh_balance()` (which is a rather expensive call against the exchange when running in live mode). 

Let's do an example:
Let's assume you got 20 pairs in your whitelist.
With this change, the bot calls the excange 20 times to get the latest balance, even if no pair generates a buy signal.
without this change, the exchange is only "asked" for pairs with active buy signal. 
Now in theory, all 20 pairs could generate a buy-signal - but still the amount of calls to `fetch_balance()` will be limited to `max_open_trades()`.

So currently, the worst case `fetch_balance()` is called is `len(whitelist)` - but with this PR it's always called `len(whitelist)` (assuming no pair is generating buy signals).

This will lead to blocks on the exchange VERY soon ... and slow the overall operation / loop duration of the bot down.
This call becomes breaking for Kraken - which only provides a "half-hearted" `fetch_balance()` method and requires an additional, second call to `fetch_open_orders()` to get the correct balances (you can look at the implementation details in `kraken.py`.

Unfortunately this can only be tested in real mode (dry-run does `refresh_balance()` locally) - so the slowdown won't be noticed in that case.

Reply:
@yazeed I mean assume custom populate_* method(s) rely they are called for each candles (for instance, aggregate or collect some data, adding one candle, with initial 'warm-up' period)... In case of your optimized code, 'warm-up' can happen at the middle of the bot operation.

Plus, with the current behavior it's easy to explain in the docs and to the users asking when the populate_* methods are called.

Reply:
Now don't ask me why my post shows up twice ... maybe answering to a previous review as part of a review has that effect ... makes it quite ... odd to see though.